pkg/agent/logrouter/logrouter.go

```diff
@@ -0,0 +1,203 @@
+//	                                  |
+//	                                  +-----> filter ---> persisted ring
+//	                                         by level
+//
+// While several destinations are passed in, these fulfill different purposes:
+//  1. System Logs: This is meant for logs to be read be the end user. While end users can read the full debug.json,
+//     system logs are meant to be more meaningful. (Note that the filter is applied upstream, and passed
+//     a on initialization. This allows the command line flags to work)
+//  2. debug.json: This is meant as a comprehensive log, suitable for sending to support
+//  3. memory ring: This is meant to hold all recent logs. Because the debug level can be noisy
+//  4. persisted ring: This handles recent info+ logs. It is persisted to the database, and can be queried
+
+package logrouter
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	agenttypes "github.com/kolide/launcher/pkg/agent/types"
+	"github.com/kolide/launcher/pkg/log/locallogger"
+	"github.com/kolide/launcher/pkg/log/ringlogger"
+	"github.com/kolide/launcher/pkg/log/teelogger"
+	"github.com/kolide/launcher/pkg/persistentring"
+)
+
+type teeloggerInt interface {
+	Add(loggers ...log.Logger)
+	Log(keyvals ...interface{}) error
+}
+
+type storeInt interface {
+	agenttypes.GetterSetterDeleterIterator
+}
+
+type ringLoggerInt interface {
+	Log(keyvals ...interface{}) error
+	GetAll() ([]map[string]any, error)
+}
+
+type logRouter struct {
+	loggerTee   teeloggerInt
+	systemTee   teeloggerInt
+	logRing     ringLoggerInt // Holds the super set of logs, in memory
+	persistRing ringLoggerInt // Only info and higher, persisted to a store
+}
+
+func New(systemLogger log.Logger) (*logRouter, error) {
+	// We always have exactly one in memory ring logger. This receives _all_ the logs. It's used to replay into new
+	// desinatations, and is expected to be a query target for recent logs
+	logRing, err := ringlogger.New(persistentring.NewInMemory(1000))
+	if err != nil {
+		return nil, fmt.Errorf("creating memory ringlogger: %w", err)
+	}
+
+	// Setup the internal logger, this rolls up all the internal destinations
+	loggerTee := teelogger.New(logRing)
+
+	// the systemLogger should copy anything it generates into the regular logger.
+	systemTee := teelogger.New(systemLogger, loggerTee)
+
+	return &logRouter{
+		loggerTee: loggerTee,
+		systemTee: systemTee,
+		logRing:   logRing,
+	}, nil
+}
+
+func (lr *logRouter) AddLogger(logger log.Logger) {
+	lr.loggerTee.Add(logger)
+}
+
+func (lr *logRouter) ReplaceSystemLogger(newSystemLogger log.Logger) {
+	newSystemTee := teelogger.New(newSystemLogger, lr.loggerTee)
+	lr.systemTee = newSystemTee
+}
+
+func (lr *logRouter) Log(keyvals ...interface{}) error {
+	return lr.loggerTee.Log(keyvals...)
+}
+
+func (lr *logRouter) Logger() log.Logger {
+	return lr.loggerTee
+}
+
+func (lr *logRouter) SystemLogger() log.Logger {
```

Comment:
How does the system logger know to write to the system log? Or rather, how do the other logs know not to write to the system log? I just want to understand how launcher make that distinction.

Reply:
Somewhat -- it doesn't. 

There are two ingestion points, Logger, and SystemLogger.

SystemLogger is a tee that routes to Logger and to the provided system log facility. There may be filtering on the system log facility. 

The caller needs to decide whether to call Logger or SystemLogger. And practically speaking, I think nothing outside of main will call SystemLogger.